enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,290 @@
+```
+
+Once the `annotations.yaml` is created with the key/value above, the following resources should appear in the operator bundle:
+
+- A service whose name matches the `ServiceName` in your annotation which routes to the readiness endpoint.
+- A deployment
+  - Whose name matches the DeploymentName used in your annotation
+  - Has a `podTemplate` with an annotation that is selected by your label
+
+### Risks and Mitigations
+
+- Not all operators may take advanatage of this feature. If an operator does not opt-in to this feature, OLM will fallback on existing behavior when checking if an operator can be upgraded.
+
+- Operator may opt-in to this feature but never update the endpoint as "upgradeable". In these situations, cluster admins must uninstall the operator and upgrade the operator manually.
+
+- If OLM reports metrics to Telemeter regarding whether or not an operator is upgradeable an unbounded metric is being reported. An unbounded metric may overwhelm Telemeter.
+
+## Design Details
+
+### Test Plan
+
+OLM's e2e testing suite will be upgraded to handle the usecase:
+
+- If a CSV defines an upgradeReadiness Service without a spec that does not exist, upgrades for that operator are blocked.
+- If a CSVless bundle defines an upgradeReadiness Service that does not exist, the bundle fails to build.
+- If a CSV defines an upgradeReadiness Service with a spec, changes to the service are reverted by OLM.
+- If a CSV/CSVless bundle defines an upgradeReadiness Service that returns a 0, upgrades to the operator are blocked. The operator then upgrades when the service is no longer blocked.
+
+Additionally, tests that use CSVs that omit the upgradeReadiness Service must continue to adhere to existing behavior.
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
+## Alternatives
+
+The are viable alternatives to the solution proposed in this enhancement.
+
+### Subsbitute the UpgradeReadinessService with an UpdateProbe
+
+It is possible that we could take advantage of the standard defined by [Liveness, Readiness, and StartUp Probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) by importing the [Prober](https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/kubelet/prober/prober.go#L48) code. There are cons and benefits to this approach defined below:
+
+**This seems to be the most viable alternative.**
+
+#### Pros
+
+- Follows a well defined probe format which supports [HTTP Gets, EXEC, and TCPSocket](https://github.com/kubernetes/api/blob/6f652b6ce59c386f4a431eb031d0339620aaff5e/core/v1/types.go#L2279-L2292) options.
+- Enables OLM to identify which POD is reporting an upgradeReadiness failure, as a service could route to one of any number of pods.
+
+#### Cons
+- Requires OLM create an import and maintain prober code which executes the probe.
+
+### Create an API so Operators may report Status
```

Comment:
to me this seems the most robust solution, one idea I had was to have operators implement a gRPC interface that OLM could then invoke, that allows a wide range of operator logic to be implemented within the operator and in any language.